### ISSUE TITLE
Decorator for computed test cases

### ISSUE THREAD

**Issue by Nysosis** [alsatian-test/alsatian#413]
Hi,

Recently came across the desired to be able to specify multiple test cases programmatically (i.e. I had a matrix of combinations I wanted to feed through).

The current `@TestCase` decorator works great, but AFAIK there's no way to easily specify multiple in that way (there could be, my knowledge of how decorators works is limited).

So what I did was to create a local `@TestCases` decorator, which was inspired heavily by the `@TestCase` decorator - except it took an `any[][]` instead of `any[]`, which allowed me to add multiple items to the `TEST_CASES` metadata at once.

This allowed me to do something like:

```
const cases = Array.from(Array(10)).map((_, i) => [i, i * 2])

// ...

@TestCase(...cases)
public FirstIsSecondTimeTwo(first: number, second: number) {
    Expect(second).toBe(first * 2)
}

```

Obviously a contrived example, but I've found similar functionality very helpful in other unit testing frameworks.

If there's a way of doing it already (native TS/JS or through the library) then point me how and we can close this issue down, otherwise I can either submit a pull request with the decorator I've done and/or we can discuss the interface for specifying computed test cases

**Comment by jamesrichford** [comment]
Nice idea @Nysosis,

Might be nice to wrap this up as a decorator that takes a generator specifically for this. What do you think?

**Comment by Nysosis** [comment]
Yup, I did consider making it use a generator at the time. Built it taking
an array for simplicity sake just to get it done quick.

I'll take a look at re-impl with a generator and submit a PR for it

**Comment by jamesrichford** [comment]
I guess if we make it a separate decorator e.g. ```TestCasesFor``` or something along those lines there's no reason it couldn't take both? So you can work with whatever suits the situation?

**Comment by Nysosis** [comment]
Well taking a quick look, I can make it work both ways, generators or
explicit array, for the one function easily enough, with a signature of
`export function TestCases(caseArguments: IterableIterator<any> | any[][])
{ ... }`. Though can't use the spread operator on on the `any[][]` argument
to make that mode a bit nice to work with (by saving you typing `[` and `]`)

**Comment by jamesrichford** [comment]
Closed by #414
